synapse/handlers/auth.py

```diff
@@ -962,6 +993,53 @@ def _do_validate_hash():
         else:
             return defer.succeed(False)
 
+    def start_sso_ui_auth(self, redirect_url: str) -> str:
+        """
+        Get the HTML for the SSO redirect confirmation page.
+
+        :param redirect_url: The URL to redirect to the SSO provider.
+        :return: The HTML to render.
+        """
+        return self._sso_auth_confirm_template.render(redirect_url=redirect_url,)
+
+    def complete_sso_ui_auth(
+        self,
+        registered_user_id: str,
+        session_id: str,
+        request: SynapseRequest,
+        requester: Requester,
+    ):
+        """Having figured out a mxid for this user, complete the HTTP request
+
+        Args:
+            registered_user_id: The registered user ID to complete SSO login for.
+            request: The request to complete.
+            client_redirect_url: The URL to which to redirect the user at the end of the
+                process.
+        """
+        # Mark the stage of the authentication as successful.
+        sess = self._get_session_info(session_id)
+        if "creds" not in sess:
+            sess["creds"] = {}
+        creds = sess["creds"]
+
+        creds[LoginType.SSO] = True
+        self._save_session(sess)
+
+        # If the user ID of the SAML session does not match the user from the
+        # request, something went wrong.
+        if registered_user_id != requester.user.to_string():
```

Comment:
shouldn't this test happen before `creds[LoginType.SSO] = True` ?

Reply:
Oops, yes. 👍

Reply:
See d3313bffa77dde3ee4ef2c0f7ef454387ec6c8c4